File: scripts/02_eda.py
import pandas as pd

print("Loading data...")
encounters = pd.read_csv("data/encounters.csv")
patients = pd.read_csv("data/patients.csv")
diagnosis = pd.read_csv("data/diagnosis.csv")
social_determinants = pd.read_csv("data/social_determinants.csv")

for df in [encounters, patients, diagnosis, social_determinants]:
    df.columns = [col.lower() for col in df.columns]

print("\n========== ENCOUNTERS ==========")
print(f"Shape: {len(encounters)} rows, {len(encounters.columns)} cols")
print("\nEncounter types (top 10):")
print(encounters['type'].value_counts().head(10))
print("\nBy admission year:")
print(encounters['admityear'].value_counts().sort_index())

print("\n========== PATIENTS ==========")
print(f"Shape: {len(patients)} rows, {len(patients.columns)} cols")
print("\nRace distribution (top 10):")
print(patients['firstrace'].value_counts().head(10))
print("\nMyChart status:")
print(patients['mychartstatus'].value_counts())

print("\n========== DIAGNOSIS ==========")
print(f"Total diagnoses: {len(diagnosis)}")
print("\nTop 25 diagnosis groups by frequency in encounters:")
enc_diag = encounters.merge(
    diagnosis[['diagnosiskey', 'groupcode', 'groupname']],
    left_on='primarydiagnosiskey',
    right_on='diagnosiskey',
    how='left'
)
print(enc_diag.groupby(['groupcode', 'groupname']).size().sort_values(ascending=False).head(25).to_frame(name='count'))

print("\n========== SOCIAL DETERMINANTS ==========")
print(f"Shape: {len(social_determinants)} rows")
print("\nDomains:")
print(social_determinants['domain'].value_counts())

n_patients_with_sdoh = social_determinants['patientdurablekey'].nunique()
pct_coverage = round(100 * n_patients_with_sdoh / len(patients), 2)
print(f"\nPatient coverage: {n_patients_with_sdoh} of {len(patients)} patients = {pct_coverage}%")

File: scripts/02_eda.R
library(tidyverse)
library(arrow)
library(janitor)

source("scripts/01_load_data.R")

cat("\n========== ENCOUNTERS ==========\n")
cat("Shape:", nrow(encounters |> select(1) |> collect()), "rows,", ncol(encounters |> head(1) |> collect()), "cols\n")
cat("Missingness (%):\n")
encounters |>
  select(1:10) |>
  summarise(across(everything(), ~ mean(is.na(.)))) |>
  collect() |>
  pivot_longer(everything(), names_to = "column", values_to = "pct_missing") |>
  arrange(desc(pct_missing)) |>
  print()

cat("\nEncounter types (Type):\n")
encounters |>
  group_by(type) |>
  count() |>
  arrange(desc(n)) |>
  collect() |>
  print()

cat("\nBy admission year:\n")
encounters |>
  group_by(admit_year) |>
  count() |>
  collect() |>
  print()

cat("\n========== PATIENTS ==========\n")
cat("Shape:", nrow(patients), "rows,", ncol(patients), "cols\n")
cat("Missingness (%):\n")
patients |>
  summarise(across(everything(), ~ mean(is.na(.)))) |>
  pivot_longer(everything(), names_to = "column", values_to = "pct_missing") |>
  arrange(desc(pct_missing)) |>
  print()

cat("\nRace distribution:\n")
patients |>
  count(first_race, sort = TRUE) |>
  print()

cat("\nMyChart status:\n")
patients |>
  count(my_chart_status, sort = TRUE) |>
  print()

cat("\n========== DIAGNOSIS ==========\n")
cat("Shape:", nrow(diagnosis), "rows\n")
cat("Top 25 diagnosis groups by usage in encounters:\n")
encounters |>
  left_join(
    diagnosis |> select(diagnosis_key, group_code, group_name),
    by = c("primary_diagnosis_key" = "diagnosis_key")
  ) |>
  group_by(group_code, group_name) |>
  count() |>
  arrange(desc(n)) |>
  head(25) |>
  collect() |>
  print()

cat("\n========== SOCIAL DETERMINANTS ==========\n")
cat("Shape:", nrow(social_determinants |> select(1) |> collect()), "rows\n")
cat("Domains:\n")
social_determinants |>
  group_by(domain) |>
  count() |>
  arrange(desc(n)) |>
  collect() |>
  print()

cat("\nPatient coverage (% with any SDoH response):\n")
n_patients_with_sdoh <- social_determinants |>
  distinct(patient_durable_key) |>
  collect() |>
  nrow()
pct_coverage <- round(100 * n_patients_with_sdoh / nrow(patients), 2)
cat(n_patients_with_sdoh, "of", nrow(patients), "patients =", pct_coverage, "%\n")
